### PR TITLE
Disable frequency-based noops via service config - part 1

### PR DESCRIPTION
**Pull request by andre4i**
[260](https://dev.azure.com/fluidframework/internal/_workitems/edit/260)

- part 1 - container reads config from service options, then from loader options
- part 2 - remove config from loader options
- part 3 - cleanup back-compat

### CODE REVIEW

File: common/lib/protocol-definitions/src/config.ts [hunk 1]
```diff
@@ -34,4 +34,21 @@ export interface IClientConfiguration {
 
     // Summary algorithm configuration. This is sent to clients when they connect
     summary: ISummaryConfiguration;
+
+    /**
+     * noopTimeFrequency & noopCountFrequency control how often a client with "write" connection needs to send
+     * noop messages in case no other ops are being sent. Any op (including noops) result in client
```
Comment (vladsud) [inline common/lib/protocol-definitions/src/config.ts:40]: Worth pointing out that noops are sent only if client did not have a chance to communicate its reference sequence number via other ops

File: packages/loader/container-loader/src/collabWindowTracker.ts [hunk 2]
```diff
@@ -28,35 +31,28 @@ import { isSystemMessage } from "@fluidframework/protocol-base";
...
+                }
+            });
+        }
     }
 
     /**
      * Schedules as ack to the server to update the reference sequence number
      */
     public scheduleSequenceNumberUpdate(message: ISequencedDocumentMessage, immediateNoOp: boolean): void {
-        // Exit early for inactive (not in quorum or not writers) clients.
```
Comment (vladsud) [inline packages/loader/container-loader/src/collabWindowTracker.ts:54]: Is this removal intentional? Looks like we can get here with non-active connection, which may cause noops to be sent, right?

**Review by vladsud — APPROVED** [review]

File: common/lib/container-definitions/src/loader.ts [hunk 3]
```diff
@@ -381,18 +381,28 @@ export type ILoaderOptions = {
...
 
     /**
      * Max time(in ms) container will wait for a leave message of a disconnected client.
     */
     maxClientLeaveWaitTime?: number;
+
+    /**
+     * Enable adding the protocol tree to the summary.
+     */
+    summarizeProtocolTree?: boolean;
```
Comment (anthony-murphy) [inline common/lib/container-definitions/src/loader.ts:405]: can you remove this. it's better as a config than an option.

File: packages/loader/container-loader/src/loader.ts [hunk 4]
```diff
@@ -128,7 +128,8 @@ function createCachedResolver(resolver: IUrlResolver) {
     return cacheResolver;
 }
 
-export interface ILoaderOptions extends ILoaderOptions1{
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/loader.ts:131]: my goal was actually the opposite here. i think we should be moving all the config here, rather than container definitions, as i don't think there is a need for loader configs to be in container definitions as they are not a layer crossing interface.

Comment (andre4i) [reply packages/loader/container-loader/src/loader.ts:131]: I'll revert this, it seems orthogonal to the work

File: packages/loader/container-loader/src/container.ts [hunk 5]
```diff
@@ -1733,13 +1728,53 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
     }
 
+    /**
+     *  #260 (ADO)
+     * back-compat: noopTimeFrequency & noopCountFrequency properties were added to
+     * IClientConfiguration in 0.59.3000. During the integration, we must read the
+     * available configuration from the loader options.
+     */
+    private getNoopConfig(): [number | undefined, number | undefined] {
+        const serviceConfiguration = this.serviceConfiguration as any;
```
Comment (NicholasCouri) [inline packages/loader/container-loader/src/container.ts:1764]: > as any [](http://example.com/codeflow?start=62&length=7)

Just so I can understand why do we need the cast as any  here ?

Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1764]: In this case, the object may or may not have the properties. When it doesn't, the compiler will complain. Currently, it will not have them, as the package needs a version bump from the definitions, where the properties were added. Afterwards, we can remove the case altogether. This is one of those (back|forward)-compat hacks which we sometimes do in order to deal with staged changes.

Comment (NicholasCouri) [reply packages/loader/container-loader/src/container.ts:1764]: Thanks - I've just realized the previous version did not have these 2 properties ;)
